src/org/labkey/test/pages/reports/ScriptReportPage.java

```diff
@@ -0,0 +1,227 @@
+    }
+
+    public static ScriptReportPage beginAtCreateReport(WebDriverWrapper webDriverWrapper, String containerPath, String reportType)
+    {
+        webDriverWrapper.beginAt(WebTestHelper.buildURL("reports", containerPath, "createScriptReport", Map.of("reportType", reportType)));
+        return new ScriptReportPage(webDriverWrapper.getDriver());
+    }
+
+    public static ScriptReportPage beginAtReport(WebDriverWrapper webDriverWrapper, String containerPath, String reportId)
+    {
+        webDriverWrapper.beginAt(WebTestHelper.buildURL("reports", containerPath, "runReport", Map.of("reportId", reportId)));
+        return new ScriptReportPage(webDriverWrapper.getDriver());
+    }
+
+    @Override
+    protected void waitForPage()
+    {
+        String activeTab = Locator.byClass("x4-tab-top-active").waitForElement(getDriver(), 10_000).getText();
+        switch (activeTab)
+        {
+            case "Report" ->
+                    waitForElement(Locator.tagWithClass("div", "reportView").notHidden().withPredicate("not(ancestor-or-self::*[contains(@class,'mask')])"), BaseWebDriverTest.WAIT_FOR_PAGE);
+            case "Source" ->
+                    shortWait().until(ExpectedConditions.visibilityOfElementLocated(Locator.byClass("reportSource")));
+        }
+    }
+
+    public ScriptReportPage clickSourceTab()
+    {
+        waitAndClick(Ext4Helper.Locators.tab("Source"));
+        waitForElement(Locator.tagWithClass("div", "reportSource").notHidden(), WAIT_FOR_PAGE);
+        return this;
+    }
+
+    public ScriptReportPage setReportSource(String source)
+    {
+        clickSourceTab();
+        setCodeEditorValue("script-report-editor", source);
+        return this;
+    }
+
+    public String saveReport(String name, boolean isSaveAs, int wait)
+    {
+        WebElement saveButton = Ext4Helper.Locators.ext4Button(isSaveAs ? "Save As" : "Save").findElement(getDriver());
+        scrollIntoView(saveButton, true);
+        clickAndWait(saveButton, wait);
+        if (null != name)
+        {
+            saveReportWithName(name, isSaveAs);
+        }
+        return getUrlParam("reportId");
+    }
+
+    /**
+     * Precondition: on save popup window
+     */
+    public void saveReportWithName(String name, boolean isSaveAs)
+    {
+        saveReportWithName(name, isSaveAs, false);
+    }
+
+    public void saveReportWithName(String name, boolean isSaveAs, boolean isExternal)
+    {
+        String windowTitle;
+        if (isExternal)
+        {
+            windowTitle = "Create New Report";
+        }
+        else
+        {
+            windowTitle = isSaveAs ? "Save Report As" : "Save Report";
+        }
+
+        Window<?> window = new Window<>(windowTitle, getWrappedDriver());
+        WebElement nameInput = Locator.xpath("//input[contains(@class, 'x4-form-field')]").findElement(window);
+        setFormElement(nameInput, name);
+        waitForFormElementToEqual(nameInput, name); // Make sure it sticks
```

Comment:
Beyond the scope of this change, but I think some of our _set_ methods should check that they did indeed do what they were asked to do.

Reply:
`setFormElement` does wait for the value to be correct but it only warns if there is a problem. There are enough places would probably fail if `setFormElement` was more strict. There are some places where the text gets immediately re-rendered into a span or something and the input element is just kind of a placeholder.